csharp/src/Ice/InputStream.cs

```diff
@@ -793,13 +792,41 @@ public object ReadSerializable()
             return result;
         }
 
+        internal static (Encoding Encoding, int Size) ReadEncapsulationHeader(
+            Encoding encoding, ReadOnlySpan<byte> buffer)
+        {
+            Debug.Assert(encoding == Encoding.V1_1 || encoding == Encoding.V2_0); // for now, only endpoints use 2.0
+
+            // With the 1.1 encoding, the encapsulation size is encoded on a 4-bytes int and not on a variable-length
+            // size, for ease of marshaling.
+            int minSize = encoding == Encoding.V1_1 ? 6 : 3;
+            if (buffer.Length < minSize)
+            {
+                throw new InvalidDataException($"encapsulation buffer has only {buffer.Length} bytes");
+            }
+            int size = ReadFixedLengthSize(encoding, buffer);
+            if (size < minSize)
+            {
+                throw new InvalidDataException($"encapsulation has only {size} bytes");
+            }
+
+            if (encoding == Encoding.V1_1 && size - 4 > buffer.Length)
+            {
+                throw new InvalidDataException(
+                    $"the encapsulation's size ({size}) extends beyond the end of the buffer");
+            }
```

Comment:
I wonder if we really need to keep these 3 checks here... I don't see what bad things could happen if it fails later down the road. Didn't we remove such underflow checks in other places such as ReadString, etc?

Reply:
Will address in a future PR.